Hot reloading/Fast refreshing throws error: "_reactNativeMmkv.getString is not a function"

I weirdly get this error **somettimes** / **often** 

context: 

I check for the token right away globally in the app, could it be that MMKV is not available then ?

<img width="325" alt="Screen Shot 2021-03-09 at 3 21 30 PM" src="https://user-images.githubusercontent.com/19273413/110476845-20268f80-80eb-11eb-9caa-b1b080892d04.png">

also the app failed building on android 

```
> Task :app:compileDebugJavaWithJavac FAILED

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.7/userguide/command_line_interface.html#sec:command_line_warnings
290 actionable tasks: 271 executed, 19 up-to-date
Note: /Users/ahmedeid/zeal/alsagheer-app-reactnative/node_modules/@react-native-community/async-storage/android/src/main/java/com/reactnativecommunity/asyncstorage/AsyncStorageModule.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: [2] Wrote GeneratedAppGlideModule with: [com.bumptech.glide.integration.okhttp3.OkHttpLibraryGlideModule, com.dylanvann.fastimage.FastImageOkHttpProgressGlideModule]
Note: /Users/ahmedeid/zeal/alsagheer-app-reactnative/node_modules/react-native-geolocation-service/android/src/main/java/com/agontuk/RNFusedLocation/LocationUtils.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
Note: /Users/ahmedeid/zeal/alsagheer-app-reactnative/node_modules/@react-native-picker/picker/android/src/main/java/com/reactnativecommunity/picker/ReactPickerManager.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
Note: /Users/ahmedeid/zeal/alsagheer-app-reactnative/node_modules/@sentry/react-native/android/src/main/java/io/sentry/react/RNSentryModule.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
/Users/ahmedeid/zeal/alsagheer-app-reactnative/android/app/src/main/java/com/alsagheer/AlsagheerJSIPackage.java:5: error: package com.reactnativemmkv does not exist
import com.reactnativemmkv.MmkvModule;
                          ^
/Users/ahmedeid/zeal/alsagheer-app-reactnative/android/app/src/main/java/com/alsagheer/MainApplication.java:44: error: cannot find symbol
            return new AlsagheerJSIPackage();
                       ^
  symbol: class AlsagheerJSIPackage
/Users/ahmedeid/zeal/alsagheer-app-reactnative/android/app/src/main/java/com/alsagheer/AlsagheerJSIPackage.java:16: error: cannot find symbol
        MmkvModule.install(jsContext, reactApplicationContext.getFilesDir().getAbsolutePath() + "/mmkv");
        ^
  symbol:   variable MmkvModule
  location: class AlsagheerJSIPackage
Note: /Users/ahmedeid/zeal/alsagheer-app-reactnative/android/app/src/debug/java/com/alsagheer/ReactNativeFlipper.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
3 errors

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:compileDebugJavaWithJavac'.
> Compilation failed; see the compiler error output for details.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 16s

error Failed to install the app. Make sure you have the Android development environment set up: https://reactnative.dev/docs/environment-setup.
Error: Command failed: ./gradlew app:installDebug -PreactNativeDevServerPort=8081
Note: /Users/ahmedeid/zeal/alsagheer-app-reactnative/node_modules/@react-native-community/async-storage/android/src/main/java/com/reactnativecommunity/asyncstorage/AsyncStorageModule.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: [2] Wrote GeneratedAppGlideModule with: [com.bumptech.glide.integration.okhttp3.OkHttpLibraryGlideModule, com.dylanvann.fastimage.FastImageOkHttpProgressGlideModule]
Note: /Users/ahmedeid/zeal/alsagheer-app-reactnative/node_modules/react-native-geolocation-service/android/src/main/java/com/agontuk/RNFusedLocation/LocationUtils.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
Note: /Users/ahmedeid/zeal/alsagheer-app-reactnative/node_modules/@react-native-picker/picker/android/src/main/java/com/reactnativecommunity/picker/ReactPickerManager.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
Note: /Users/ahmedeid/zeal/alsagheer-app-reactnative/node_modules/@sentry/react-native/android/src/main/java/io/sentry/react/RNSentryModule.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
/Users/ahmedeid/zeal/alsagheer-app-reactnative/android/app/src/main/java/com/alsagheer/AlsagheerJSIPackage.java:5: error: package com.reactnativemmkv does not exist
import com.reactnativemmkv.MmkvModule;
                          ^
/Users/ahmedeid/zeal/alsagheer-app-reactnative/android/app/src/main/java/com/alsagheer/MainApplication.java:44: error: cannot find symbol
            return new AlsagheerJSIPackage();
                       ^
  symbol: class AlsagheerJSIPackage
/Users/ahmedeid/zeal/alsagheer-app-reactnative/android/app/src/main/java/com/alsagheer/AlsagheerJSIPackage.java:16: error: cannot find symbol
        MmkvModule.install(jsContext, reactApplicationContext.getFilesDir().getAbsolutePath() + "/mmkv");
        ^
  symbol:   variable MmkvModule
  location: class AlsagheerJSIPackage
Note: /Users/ahmedeid/zeal/alsagheer-app-reactnative/android/app/src/debug/java/com/alsagheer/ReactNativeFlipper.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
3 errors

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:compileDebugJavaWithJavac'.
> Compilation failed; see the compiler error output for details.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 16s

    at makeError (/Users/ahmedeid/zeal/alsagheer-app-reactnative/node_modules/execa/index.js:174:9)
    at /Users/ahmedeid/zeal/alsagheer-app-reactnative/node_modules/execa/index.js:278:16
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async runOnAllDevices (/Users/ahmedeid/zeal/alsagheer-app-reactnative/node_modules/@react-native-community/cli-platform-android/build/commands/runAndroid/runOnAllDevices.js:94:5)
    at async Command.handleAction (/Users/ahmedeid/zeal/alsagheer-app-reactnative/node_modules/@react-native-community/cli/build/index.js:186:9)
info Run CLI with --verbose flag for more details.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

--- a/ios/Mmkv.mm
+++ b/ios/Mmkv.mm
@@ -115,18 +115,26 @@ static void install(jsi::Runtime & jsiRuntime)
     jsiRuntime.global().setProperty(jsiRuntime, "mmkvGetAllKeys", std::move(mmkvGetAllKeys));
 }
 
-- (void)setBridge:(RCTBridge *)bridge
+- (void)setup
 {
-    _bridge = bridge;
-    _setBridgeOnMainQueue = RCTIsMainQueue();
-
     RCTCxxBridge *cxxBridge = (RCTCxxBridge *)self.bridge;
     if (!cxxBridge.runtime) {
+        // retry 10ms later
+        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 0.001 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
+            [self setup];
+        });
         return;
     }
 
     [MMKV initializeMMKV:nil];
     install(*(jsi::Runtime *)cxxBridge.runtime);
+}
+
+- (void)setBridge:(RCTBridge *)bridge
+{
+    _bridge = bridge;
+    _setBridgeOnMainQueue = RCTIsMainQueue();
+    [self setup];
 }
 
 - (void)invalidate {

--- a/ios/Mmkv.mm
+++ b/ios/Mmkv.mm
@@ -119,7 +119,7 @@ static void install(jsi::Runtime & jsiRuntime)
 {
     RCTCxxBridge *cxxBridge = (RCTCxxBridge *)self.bridge;
     if (!cxxBridge.runtime) {
-        // retry 10ms later
+        // retry 10ms later - THIS IS A WACK WORKAROUND. wait for TurboModules to land.
         dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 0.001 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
             [self setup];
         });

